### PR TITLE
Include tool for browsing model/submodel structure

### PR DESCRIPTION
**Pull request by jeremy9959**
This PR is the second attempt to address #9654.

1) I added a file  ```bokeh/bokeh/util/structure.py``` that defines a class ```BokehStructureGraph```
which, from a bokeh model, offers:

- a networkx graph of the model/submodel structure
- a pandas dataframe of all the attributes for all the submodels, with their types, values, and docstrings.
- a bokeh model that renders the networkx graph.  Clicking on a node (submodel) reveals the
attributes of that submodel, and their values.  This can be used in a jupyter notebook as it relies on JS callbacks.

2) I created a subdirectory ```bokeh/examples/tools``` with three files:
- ```structure_graph.py``` which illustrates the structure graph
-```ModelStructureExample.ipynb``` which demonstrates the class in a jupyter notebook
- a README

### CODE REVIEW

File: bokehjs/package-lock.json [hunk 1]
```diff
@@ -3008,7 +3008,7 @@
     },
     "os-tmpdir": {
       "version": "1.0.2",
-      "resolved": "https://registry.npmjs.org/os-tmpdir/-/os-tmpdir-1.0.2.tgz",
+      "resolved": "http://registry.npmjs.org/os-tmpdir/-/os-tmpdir-1.0.2.tgz",
```
Comment (mattpap) [inline bokehjs/package-lock.json]: Please revert this change. This is a side effect of npm being dubious piece of software. I suggest using `git add -p` (or `--patch`) to exclude unwanted changes.

Comment (jeremy9959) [reply bokehjs/package-lock.json]: I hope I've succeeded.  I don't fully understand it but I definitely didn't want it there.

Comment (jeremy9959) [reply bokehjs/package-lock.json]: Looks like it's gone now.  I'll read up on ```git add -p```.

**Comment by jeremy9959** [comment]
Hoping the windows test failure is an artifact, I don't have a windows machine to look at it.

**Comment by mattpap** [comment]
> Hoping the windows test failure is an artifact, I don't have a windows machine to look at it.

Yeah, it's unrelated.

File: bokeh/util/structure.py [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+   works in a jupyter notebook.
+
+   Suppose M is a bokeh model (i.e. a plot or figure).
+
+      - K = BokehStructureGraph(M) creates the graph, which is itself a bokeh model.
+      - K.show() calls show on the model to display it.
+      - K.model returns the model consisting of the structure graph and associated datatable.
+      - K.graph returns the directed graph in networkx format
+      - K.property_df returns the dataframe of models, submodels and their properties.
+"""
```
Comment (bryevdv) [inline bokeh/util/structure.py:13]: Docstrings (here and throughout) should be updated to match existing project style:
```
""" Short description with a space in front of the sentence.

Further description. Not indented (aligned with docstring quotes).

Empty line at the end.

"""
```

File: bokeh/util/structure.py [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+
+# Standard library imports
+from itertools import permutations
+
+# External imports
+import networkx as nx
+import pandas as pd
```
Comment (bryevdv) [inline bokeh/util/structure.py:20]: These are not hard dependencies of the project. It would be better for them to be protected imports inside any functions that use them, rather than at the top level.

File: bokeh/util/structure.py [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+    ResetTool,
+    TableColumn,
+    TapTool,
+)
+
+# -----------------------------------------------------------------------------
+# General API
+# -----------------------------------------------------------------------------
+
+__all__ = ['BokehStructureGraph']
```
Comment (bryevdv) [inline bokeh/util/structure.py:49]: prefer tuple to list here

File: bokeh/util/structure.py [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+                prop_source=self._prop_source,
+                f=self._groupfilter,
+                table=self._data_table,
+            ),
+            code=js_code,
+        )
+
+        self._node_source.selected.js_on_change("indices", js)
+        layout = column(self._graph_plot, self._data_table)
+        return layout
```
Comment (bryevdv) [inline bokeh/util/structure.py:315]: This module needs at least minimal tests to exercise the public API (so that it can't change unintentionally)

File: bokeh/util/structure.py [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+                keep_edge = False
+                for y in H.properties():
+                    if test_condition(s, y, H):
+                        keep_edge = True
+                if not keep_edge:
+                    dead_edges.append((id, x))
+        K.remove_edges_from(dead_edges)
+        return K
+
+    def obj_props_to_df2(self, obj):
```
Comment (bryevdv) [inline bokeh/util/structure.py:157]: does this need to be public API?

File: examples/tools/structure_graph.py [hunk 7]
```diff
@@ -0,0 +1,10 @@
+from bokeh.io import show
+from bokeh.layouts import column
+from bokeh.plotting import figure
+from bokeh.util import structure
+
+# draw the structure graph of a basic figure model
+f = figure(width=400,height=400)
+f.line(x=[1,2,3],y=[1,2,3])
+K = structure.BokehStructureGraph(f)
+show(column(f,K.model))
```
Comment (bryevdv) [inline examples/tools/structure_graph.py:10]: I'd suggest making a free function that is the "common" public API, to simplify this. It would just create a `BokehStructureGraph` for you and return the model, as a convenience. Something like:
```
from bokeh.util.structure import generate_structure_plot

sg = generate_structure_plot(f)

show(column(f, sg))
```

If there are not actually any useful configurable things on `BokehStructureGraph` right now then I would even say it could be private `_BokehStructureGraph` for now. That would make the fewer API commitments now, which I think is good (more flexible later)

File: examples/tools/README.md [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+   and docstrings.
+   - a bokeh model that draws the graph of models and submodels; clicking on a node reveals the attributes
+   of that submodel and their values.
+
+    See [this jupyter notebook](./ModelStructureExample.ipynb)
+    or run:
+
+    ```
+    $ python structure_graph.py
+    ```
```
Comment (bryevdv) [inline examples/tools/README.md:19]: I'd prefer to avoid an entire new subdirectory at the top level. Suggest moving to `examples/howto/structure/`

File: bokeh/util/structure.py (hunk 5 above)
Comment (bryevdv) [reply bokeh/util/structure.py:315]: See also comment on the example about API

**Comment by bryevdv** [comment]
Also there have been some testing/infra changes since March, this should be merged/rebased against current master.

**Comment by jeremy9959** [comment]
I’ll take a look. I’m still interested.

Jeremy Teitelbaum
UConn Math

________________________________

File: examples/tools/README.md (hunk 8 above)
Comment (jeremy9959) [reply examples/tools/README.md:19]: Done.

File: bokeh/util/structure.py (hunk 4 above)
Comment (jeremy9959) [reply bokeh/util/structure.py:49]: Done.

File: examples/tools/structure_graph.py (hunk 7 above)
Comment (jeremy9959) [reply examples/tools/structure_graph.py:10]: Done

File: bokeh/util/structure.py (hunk 6 above)
Comment (jeremy9959) [reply bokeh/util/structure.py:157]: No, I've made the whole class and its methods private.  Actually this function can be useful generally -- it organizes all the properties and their docstrings in a dataframe -- but for now it can be private.

File: bokeh/util/structure.py (hunk 3 above)
Comment (jeremy9959) [reply bokeh/util/structure.py:20]: I've done this but I'm not sure I agree since although only certain functions use them, they are pretty fundamental to the way I've put this together so to do anything useful they are going to be imported.

File: bokeh/util/structure.py (hunk 2 above)
Comment (jeremy9959) [reply bokeh/util/structure.py:13]: Done

File: bokeh/util/structure.py (hunk 5 above)
Comment (jeremy9959) [reply bokeh/util/structure.py:359]: @bryevdv I could use some advice on the right test.  I have a test that compares the submodel graph computed by the code against a known example, but that is based on the private API and it's more focused on the internals of the algorithm. The current setup has the public API returning a model.  I could precompute a structure graph for a simple model and have the test compare the result to this precomputed example -- is that what you have in mind?

Comment (bryevdv) [reply bokeh/util/structure.py:359]: @jeremy9959  My comment was about the minmum bar, which is just:
> Any test that calls the function at all (so that if the function was removed or deleted unintentionally, that test would fail)
A easy next step would be to look at a return result and make sure it is the expected type (and might do so for several inputs to test corner cases)

More thorough would be what you describe, to actually compare return values. If that is something you can do, it'd definitely encourage that level of thoroughness.

Comment (jeremy9959) [reply bokeh/util/structure.py:359]: I put in a sort of compromise.  I compute the graph for a simple model and see if
it has the right number of submodels. See what you think.

Comment (jeremy9959) [reply bokeh/util/structure.py:359]: However, this reveals a new problem.  The networkx graph layout I am using needs pydot, and that doesn't seem to be available in the CI environment.

Comment (jeremy9959) [reply bokeh/util/structure.py:359]: And, similarly, pandas is shown as "module not found" in the tests.

Comment (bryevdv) [reply bokeh/util/structure.py:359]: is the version of `pydot` you need available via conda? Then the `environment-test*.yml` files under the `ci` directory can be updated. If the test needs pandas then the pytest fixture `pd` should be added to the test arguments so that the test is skipped on the "minimal" test run (which specifically tests that basic functions work even if things like pandas are not installed)

For `isort` and general linter issues I def recommend running `pytest -m tests/codebase` before a push

Comment (jeremy9959) [reply bokeh/util/structure.py:359]: OK, I think I have it straightened out.

**Comment by bryevdv** [comment]
@jeremy9959 awesome all green! I think just two more small changes:

* There is a top-level `environment.yml` that should be updated to match
* A `tests/examples.log`file got checked in by accident (should be deleted)

But I also wanted to ask your thoughts about one other change (sorry it did not occur ealier): What do you think about moving to `bokeh.models.utils`? Or, alternatively, importing into `bokeh.models` directly to be exposed publicly from there? he reason I ask is because it's pretty unusual for users to use anything out of `bokeh.utils` directly, so maybe there is a better place for this to call "home".

**Comment by jeremy9959** [comment]
@bryevdv wrote:

> * There is a top-level `environment.yml` that should be updated to match

 -- done
> * A `tests/examples.log`file got checked in by accident (should be deleted)

-- done

>  But I also wanted to ask your thoughts about one other change (sorry it did not occur earlier): What do you think about moving to `bokeh.models.utils`? Or, alternatively, importing into `bokeh.models` directly to be exposed publicly from there? he reason I ask is because it's pretty unusual for users to use anything out of `bokeh.utils` directly, so maybe there is a better place for this function to call "home".

I'd defer to your judgement here.  I'm not sure how a user would come across this function
in the normal course of events, and I guess `bokeh.models` is pretty much center stage.  I'm happy to make the move if you think it's a good idea.

**Comment by bryevdv** [comment]
@mattpap @philippjfr do either of you have thoughts/prefs on where to put this funtion?

**Comment by philippjfr** [comment]
No strong preference, but `bokeh.models.utils` seems fine to me.

**Comment by bryevdv** [comment]
OK that sounds good, @jeremy9959 please move `bokeh.models.utils` and we will get this merged!

**Comment by jeremy9959** [comment]
@bryevdv  just so I do this right.  Are you thinking

```
from bokeh.models.utils import generate_structure_graph
```
or
```
from bokeh.models.utils.structure import generate_structure_graph
```

Also consistency suggests `util` not `utils`.

**Comment by bryevdv** [comment]
@jeremy9959 Let's keep things simple with 
```
from bokeh.models.util import generate_structure_graph
```
The `bokeh.util` package is very broad and vague in scope but presumably `bokeh.models.util` has a more narrow focus so that one module will suffice .

**Comment by bryevdv** [comment]
Def recommend installing a git hook (pre-commit or at least pre-push) that runs the codebase checks :)

**Comment by bryevdv** [comment]
@jeremy9959 Thanks for all your patience pushing this through. FYI we are in the process of changing how re branch and release. This will be merged to master so it will show up in a 2.2 release in ~4 weeks (but not in the 2.1.1 patch release coming next week)

**Comment by jeremy9959** [comment]
Thank YOU for your patience!  This has been incredibly educational for me and I'm honored to have some code as a tiny part of the bokeh distribution!
